src/java/org/apache/cassandra/db/compaction/unified/CostsCalculator.java

```diff
@@ -0,0 +1,279 @@
+    public void close()
+    {
+        writeLock.lock();
+
+        try
+        {
+            logger.debug("Stopping cost calculations for {}", strategy.getMetadata());
+            future.cancel(false);
+            logger.debug("Stopped cost calculations for {}", strategy.getMetadata());
+        }
+        finally
+        {
+            writeLock.unlock();
+        }
+    }
+
+    @VisibleForTesting
+    void sampleValues()
+    {
+        writeLock.lock();
+
+        try
+        {
+            partitionsRead.update(env.partitionsRead());
+            bytesInserted.update(env.bytesInserted());
+
+            numSSTables.update(strategy.getSSTables().size());
+            spaceUsed.update(strategy.getSSTables().stream().map(SSTableReader::onDiskLength).reduce(0L, Long::sum));
+        }
+        catch (Throwable err)
+        {
+            JVMStabilityInspector.inspectThrowable(err);
+            logger.error("Failed to update values: {}/{}", err.getClass().getName(), err.getMessage(), err);
+        }
+        finally
+        {
+            writeLock.unlock();
+        }
+    }
+
+    /**
+     * @return the estimated read cost for the given number of partitions, in milliseconds
+     */
+    private double getReadCost(double partitionsRead)
+    {
+        return (env.sstablePartitionReadLatencyNanos() * partitionsRead) / TimeUnit.MILLISECONDS.toNanos(1);
+    }
+
+    /**
+     * Calculate the projected read cost for user queries.
+     *
+     * The projected read cost is given by the number of partitions read, times the mean partition latency and is calculated
+     * by {@link this#getReadCost(double)}. This value is then multiplied by the number of sstables we're likely to hit
+     * per partition read and the read multiplier.
+     * <p/>
+     * The number of sstables is calculated as Math.min(1 + env.bloomFilterFpRatio() * RA / survivalFactor, RA). Here we
+     * assume there is going to be at least one sstable accessed, possibly more in case of :
+     *
+     * - bloom filter's false positives;
+     * - partitions not surviving a compaction (1/survivalFactor is the limit of the sum of (1-survivalFactor)^n), that
+     *   is partitions that would not exist if compaction was done; Note that the survival factor is currently fixed to 1.
+     *
+     * The RA is then a cap since we cannot read more than RA sstables, which are the sstables that exist because
+     * compactions allows them to exist.
+     * </p>
+     * The read multiplier is a factor that operators can use to tweak the algorithm.
+     * </p>
+     * @param RA the expected read amplification due to the current choice of compaction strategy
+     *
+     * @return the projected read cost for user queries
+     */
+    public double getReadCostForQueries(int RA)
+    {
+        readLock.lock();
+
+        try
+        {
+            return getReadCost(partitionsRead.avg.get()) * Math.min(1 + env.bloomFilterFpRatio() * RA / survivalFactor, RA) * readMultiplier;
```

Comment:
Is the survival factor useful in any way currently? If not, let's remove all references for now.

Reply:
It is not used in this patch, but cndb uses it to handle replicated data on L0. Therefore, I'd prefer to keep it.